File: project_euler_unified/problem_097/solution.py
"""
Project Euler Problem 97: Large Non-Mersenne Prime

Find the last ten digits of 28433 * 2^7830457 + 1.

Uses Python's built-in three-argument pow(base, exp, mod) which implements
binary exponentiation (repeated squaring) in O(log exp) modular multiplications.

Answer: 8739992577
"""


def main():
    mod = 10**10
    result = (28433 * pow(2, 7830457, mod) + 1) % mod
    print(result)


if __name__ == "__main__":
    main()

File: project_euler_unified/problem_097/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    // Compute (28433 * 2^7830457 + 1) mod 10^10
    // Using binary exponentiation with __int128 to avoid overflow.

    const long long mod = 10000000000LL;

    auto mulmod = [](long long a, long long b, long long m) -> long long {
        return ((__int128)a * b) % m;
    };

    // Binary exponentiation: 2^7830457 mod 10^10
    long long base = 2, exp = 7830457, result = 1;
    base %= mod;
    while (exp > 0) {
        if (exp & 1)
            result = mulmod(result, base, mod);
        base = mulmod(base, base, mod);
        exp >>= 1;
    }

    result = mulmod(28433LL, result, mod);
    result = (result + 1) % mod;

    cout << result << endl;
    return 0;
}
